hazelcast/src/main/java/com/hazelcast/map/impl/operation/MapFetchIndexOperation.java

```diff
@@ -123,41 +122,90 @@ protected void runInternal() {
         }
     }
 
-    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NPathComplexity"})
+    @SuppressWarnings({"checkstyle:CyclomaticComplexity", "checkstyle:NPathComplexity", "checkstyle:MethodLength"})
     private MapFetchIndexOperationResult runInternalSorted(InternalIndex index) {
-        List<QueryableEntry<?, ?>> entries = new ArrayList<>(sizeHint + EXCESS_ENTRIES_RESERVE);
+        List<QueryableEntry<?, ?>> entries = new ArrayList<>(sizeLimit + 1);
         int partitionCount = getNodeEngine().getPartitionService().getPartitionCount();
+        DataComparator comparator = new DataComparator();
 
         for (int i = 0; i < pointers.length; i++) {
             IndexIterationPointer pointer = pointers[i];
+            Data lastEntryKeyData = pointer.getLastEntryKeyData();
+
             Iterator<IndexKeyEntries> entryIterator = getEntryIterator(index, pointer);
 
             while (entryIterator.hasNext()) {
                 IndexKeyEntries indexKeyEntries = entryIterator.next();
-                @SuppressWarnings({"unchecked", "rawtypes"})
-                Collection<QueryableEntry<?, ?>> keyEntries = (Collection) indexKeyEntries.getEntries();
-                if (partitionIdSet == null) {
-                    entries.addAll(keyEntries);
-                } else {
-                    for (QueryableEntry<?, ?> entry : keyEntries) {
+                @SuppressWarnings({"rawtypes"})
+                Iterator<QueryableEntry> keyEntries = indexKeyEntries.getEntries();
+
+                // Skip until the entry last read
+                if (lastEntryKeyData != null) {
+                    while (keyEntries.hasNext()) {
+                        QueryableEntry<?, ?> entry = keyEntries.next();
+
+                        int comparison = comparator.compare(entry.getKeyData(), lastEntryKeyData);
+                        if (isDescendingEntryKey(pointer)) {
+                            comparison = -comparison;
+                        }
+
+                        if (comparison == 0) {
+                            break;
+                        } else if (comparison > 0) {
+                            entries.add(entry);
+                            lastEntryKeyData = entry.getKeyData();
+                            break;
+                        }
```

Comment:
Micro-optimization: on the more common path until we find the entry do only one comparison:
```suggestion
                        if (comparison >= 0) {
                            if (comparison > 0) {
                                entries.add(entry);
                                lastEntryKeyData = entry.getKeyData();
                            }
                            break;
                        }
```